TACC/Core-CMS
# FP-1318: Create Section Pattern via Bootstrap Container Plugin

## To Do

- [x] [install `django-cms/djangocms-bootstrap4`](https://github.com/TACC/Core-CMS/pull/431/files#r981674483) at [v3.0.0](https://github.com/django-cms/djangocms-bootstrap4/releases/tag/3.0.0)
- [x] re-test
- ~~fix [django cms error upon adding container](https://github.com/TACC/Core-CMS/assets/62723358/d94cc4ce-c32c-4736-97c4-f8957d504f30)~~
- [x] ping reviewers

## Overview

Quickly create existing TACC Section _pattern_ via already-installed[^2] Django Bootstrap4 Container _plugin_.

 _CMS Admin/Designer has approved (via private review), but implementation needs dev review._

[^2]: The install is given a new requirement (in `requirements.txt`) to fix a minor UI bug.

## Tickets

* [FP-1318](https://jira.tacc.utexas.edu/browse/FP-1318) → [__FP-1478__](https://jira.tacc.utexas.edu/browse/FP-1478)


## Changes

To our default `settings`:
* Add `DJANGOCMS_BOOTSTRAP4_GRID_CONTAINERS` ([docs](https://github.com/django-cms/djangocms-bootstrap4#settings)).
    * to allow user to quickly create light Section or dark Section (as Container or not)
* (Optional) Load `gettext_lazy`.
    * to support translation of choice group labels and values[^1]

To our default requirements:
* Install [fork of djangocms-bootstrap4](https://github.com/tacc-wbomar/djangocms-bootstrap4) that has [approved](https://github.com/django-cms/djangocms-bootstrap4/pull/145) fix.
    * to fix minor UI bug where Container plugin has blank short description for grouped choices[^1]

[^1]: I.e. `<optgroup>` labels and `<option>` values in selection dropdown when `DJANGOCMS_BOOTSTRAP4_GRID_CONTAINERS` has nested tuples.


## Screenshots

https://user-images.githubusercontent.com/62723358/150383016-0a028625-cd86-4400-baca-d7f7e77db301.mov

## Testing

> __local__:  [instructions](https://github.com/TACC/Core-CMS/wiki/Test-CMS-Changes)
> __remote__: https://dev.cep.tacc.utexas.edu/ ([build](https://jenkins01.tacc.utexas.edu/view/WMA%20CEP/job/Core_CMS_Build/360/), [deploy](https://jenkins01.tacc.utexas.edu/job/Core_Portal_Deploy/1042/))

1. On any page, create a Container plugin.
2. Choose some of the eight CONTAINER TYPEs.
3. Confirm an element is added each time with the appropriate `class` attribute.\*

\* To see what the section should **look** like, you may compare to pattern library: [light](https://cep.tacc.utexas.edu/static/ui/components/detail/o-section--style-light.html), [dark](https://cep.tacc.utexas.edu/static/ui/components/detail/o-section--style-dark.html), [muted](https://cep.tacc.utexas.edu/static/ui/components/detail/o-section--style-muted.html).


## Notes

### Related PRs
- Use Style Plugin Instead & Define Tag Types:
    https://github.com/TACC/Core-CMS/pull/430
- Use Style Plugin Instead & Support Other Classes:
    https://github.com/TACC/Core-CMS/pull/419
